Fix tray icon on KDE5 and Gnome3

agilob opened the pull request:
On KDE5 qTox will use Qt backend instead buggy status-notifier, on KDE5, Qt 5.5 it looks like this now:

Commit fix some issues:
- SVG icons are used instead pixmaps
- transparent icon on Qt5.5 will be fixed (uses transparent SVG rendered qimage instead qpixmap)
- adds keyboard shourtcut to quit application, also KDE5 adds by default menu action row to Qt application - "Quit", so our Quit was renamed to Exit. Exit has system-default icon and system default keyboards shortcut. Difference will be visible only on plasma-shell (see screenshot) where `Quit` asks if you really want to quit application and exit just quits without asking.

Commits were tested on KDE5, Gnome3 by me and i3 by @zetok.

![qtox9](https://cloud.githubusercontent.com/assets/2544251/9562233/aeb8fc30-4e5e-11e5-8db3-3597837b6ab1.png)

Before:
![snapshot2](https://cloud.githubusercontent.com/assets/2544251/9563755/8e87d990-4e86-11e5-9726-378b6b66132a.png)

dubslow committed:
add ringtone (not well done, but still better than status quo)

agilob committed:
Merge remote-tracking branch 'upsteam/master'

Conflicts:
	src/widget/widget.h

agilob committed:
use Qt backend of KDE5

commit fixes problem with usage of GTK3 fallback menu action used on KDE5, probably to a bug in status notifier
- allows using Qt backend on KDE5 and other plasma related env.
- icon in tray menu and tray icon are loaded from svg, their size is not limited to 50px now
- plasma adds extra actionmenu to tray so quit was renamed to exit
- exit action usues system-default icon for quit and system default keyboard shourtcut to close qtox

agilob committed:
fix icon scaling on gnome3 by scaling icons down

agilob commented:
@antis81 could you test it on your desktop and merge?

antis81 commented on src/widget/systemtrayicon.h:
```diff
@@ -48,6 +49,8 @@ class SystemTrayIcon : public QObject
 private:
     SystrayBackendType backendType;
     QSystemTrayIcon* qtIcon;
+    Widget *parent;
```
Proposal: Rename to `mParent`. Thus, `this->parent = parent` would become `mParent = parent` in c'tor implementation [in cpp line 31](https://github.com/tux3/qTox/pull/2132/files#diff-1b85c272c0d41b3a9efecede73cfa60eR31).

antis81 commented on src/widget/widget.h:
```diff
@@ -160,6 +160,9 @@ private slots:
     void friendListContextMenu(const QPoint &pos);
 
+private:
+    const int icon_size = 15;
```
Please move initialization to c'tor so order stays clear. Not every "feature" c++11/14 offers is a good one ...

agilob replied:
`to c'tor`? 

what does it mean?

constructor -.-'

antis81 replied:
Yep :) - sorry, it's an older c++ name convention.

antis81 commented on src/widget/systemtrayicon.cpp:
```diff
@@ -26,8 +26,9 @@
 #include "src/persistence/settings.h"
 
-SystemTrayIcon::SystemTrayIcon()
+SystemTrayIcon::SystemTrayIcon(Widget *parent)
```
@agilob Uhm ... we seem to have a private implementation here ... so I think that ...
a) `parent` member should be moved there
b) private class should be initialized here

Right?

agilob replied:
that was used to experiment with qt implementation/status notifier, in fact is should be removed because parent is never used anyway. Thanks for spotting this o/

antis81 commented on src/widget/systemtrayicon.h:
```diff
@@ -22,6 +22,7 @@
 #define SYSTEMTRAYICON_H
 
 #include "systemtrayicon_private.h"
```
@agilob Though unrelated to your current changes in this PR. If I understand PIMPL right and not am completely blind, this line should be moved to cpp, to keep implementation **private**. :smile:

agilob committed:
remove dependency injection from systemtrayicon

as it was used as experiment only
fix compilation warning about unchecked switch case

agilob commented:
@zetok why travis didn't build first and second commit?

antis81 commented:
@agilob I still need to setup a working build environment. The qtox repo, compiler, etc. is installed. Anything I need to be aware of despite of the lib dependencies? Anyway, I'll set it up and report back. Not sure though, if I really can check the "system tray icon" as I don't have a "system tray". :smile: -> using `cairo-dock` panel replacement.

Anyone else running a common Linux distro and "system tray"?

agilob commented:
we just need to test it on unity.

antis81 commented:
I'm using (LXDE/LXQt) for working environment. Anyone using a unity environment to check, if this works?

antis81 commented on src/widget/widget.cpp:
```diff
@@ -1540,12 +1554,20 @@ QString Widget::getStatusIconPath(Status status)
-        return getStatusIconPixmap(status, w, h);
-    else
-        return QIcon(getStatusIconPath(status));
+#ifdef __linux__
```
How about `#ifdef Q_OS_LINUX` instead?

antis81 commented on src/widget/widget.cpp:
```diff
@@ -1540,12 +1554,20 @@ QString Widget::getStatusIconPath(Status status)
+    if (desktop.isEmpty())
+        desktop = getenv("DESKTOP_SESSION");
+    desktop = desktop.toLower();
+    if (desktop == "xfce" || desktop.contains("gnome") || desktop == "mate") {
```
Maybe change this to "exclude" strategy:

``` c++
if (desktop != "unity") { ... }
```

agilob replied:
not really, this if must cover window managers based on GTK2/3. `desktop!="unity"` will break support for i3, awesome and other wired managers.

antis81 replied:
Ok then, no problem.

antis81 commented on src/widget/widget.cpp:
```diff
@@ -1540,12 +1554,20 @@ QString Widget::getStatusIconPath(Status status)
-        return QIcon(getStatusIconPath(status));
+#ifdef __linux__
+    QString desktop = getenv("XDG_CURRENT_DESKTOP");
+    if (desktop.isEmpty())
```
Can you please add the `{ }` brackets? Improves readability and reduces potential for mistakes.

antis81 commented:
@agilob Apart from the minor proposals I wrote, it looks good to me. Would you mind cleaning them up?

agilob committed:
core review changes - formatting

agilob commented:
done.

zetok commented:
> @zetok why travis didn't build first and second commit?

travis builds only latest HEAD, with build triggered by push to repo/PR

ProMcTagonist commented:
Why does it have "Exit | Ctrl + Q" and "Quit" both?

agilob commented:
> Why does it have "Exit | Ctrl + Q" and "Quit" both?

Read my PR above...

ProMcTagonist commented:
It's standard to include your own on KDE?

antis81 commented:
> Why does it have "Exit | Ctrl + Q" and "Quit" both?

@agilob Maybe setting `QAction::QuitRole` helps to avoid double entry.

agilob commented:
@antis81 yes 8) 
![snapshot3](https://cloud.githubusercontent.com/assets/2544251/9564206/69c882b6-4e96-11e5-8101-d8c58f543ba7.png)

antis81 commented:
@agilob hehe. Hm ... only an idea: Maybe you can check for an existing quit action and only add, if none was found.

Btw.: This also affects OSX. It's not so obvious though, as the quit action is bound to a special menu called "application menu".

agilob committed:
add quit role to quit action in trayicon

agilob commented:
I'm not sure if OSX will be affected by those changes. I added quitrole to that action, but it doesnt change anything on KDE5 and I can't find any working example that doesn't reproduce this issue, I checked some bitcoin clients and torrent clients written in c++/qt. I'll check source code of KDE tomorrow. Isn't application menu related to menubar in qmainwindow? [That's what documentation says](http://doc.qt.io/qt-4.8/qmenubar.html#qmenubar-on-mac-os-x)

agilob committed:
add logout action to tray menu

agilob committed:
use plaintext label for username and status in header

agilob commented:
I added logout action:
![snapshot7](https://cloud.githubusercontent.com/assets/2544251/9566964/522b0562-4f12-11e5-8e01-21da3a8b9bd2.png)

Another commit partially fixes #1999, we can't completely fix 1999 but username and status will be printed as plaintext:
![snapshot8](https://cloud.githubusercontent.com/assets/2544251/9567131/ad88be5e-4f17-11e5-93ad-5077f7d74452.png)

agilob commented:
@antis81 I'm pretty sure OSX will not be affected by those changes, and if OSX would add second quit action, then it ALREADY is there, as my changes affect only KDE5 and Gnome3, nothing else. I changed used backend used on KDE5 from status notifier to qt and fixed rendering transparent tray icon. If OSX adds second quit action, it already is there, my changes won't cause it.

agilob committed:
Merge branch 'master' of github.com:tux3/qTox into styling

agilob commented:
The last commit here fixes conflict with origin/master, sorry for misleading commit message.

tux3 commented:
I'm getting a crash at systemtrayicon.cpp:287 on a KDE5 system.

tux3 commented on src/widget/systemtrayicon.cpp:
```diff
@@ -95,28 +95,12 @@ SystemTrayIcon::SystemTrayIcon()
+        qDebug() << "Using the Qt backend";
+        qtIcon = new QSystemTrayIcon;
+        backendType = SystrayBackendType::Qt;
+        connect(qtIcon, &QSystemTrayIcon::activated, this, &SystemTrayIcon::activated);
```
That seems wrong, you're using the Qt backend inside an #ifdef ENABLE_SYSTRAY_STATUSNOTIFIER_BACKEND

The ifdef should just be removed if you don't actually use status notifier here.

agilob commented:
Hmm. how did you compile it? I used it for a week now without any problems

zetok commented:
@agilob looks like it fails to build: https://build.tox.chat/job/qTox-testing_build_linux_x86-64_release/10/console

agilob committed:
systemtrayicon: remove unnecessary ifdef

agilob committed:
Merge branch 'master' of github.com:agilob/qTox into styling

agilob committed:
Merge with tux3:upstream/master

agilob commented:
Fixed and merged with upstream master.

agilob commented:
@tux3 could you try this now? Could you tell me where is the crash from?

antis81 commented on src/widget/systemtrayicon.cpp:
```diff
@@ -94,31 +94,13 @@ SystemTrayIcon::SystemTrayIcon()
     }
-    #endif
     else if (desktop == "kde"
              && getenv("KDE_SESSION_VERSION") == QString("5"))
```
Uhm, isn't this the same as previous condition above? Btw.: Better user QStringLiteral("5").

tux3 commented:
@agilob 
It works for me now, with just a couple issues.
- [ ] When I hover the mouse over the exit icon, it becomes a completely different icon:

![snapshot34](https://cloud.githubusercontent.com/assets/5155966/10362853/4607b498-6db1-11e5-8296-e1b0df5c083e.png) ![snapshot35](https://cloud.githubusercontent.com/assets/5155966/10362854/460ab922-6db1-11e5-8ab5-7ea65a05785b.png)
- As @antis81 noted, the `else if` line 104 can be removed:

```
    else if (desktop == "kde"
             && getenv("KDE_SESSION_VERSION") == QString("5"))
    {
        backendType = SystrayBackendType::KDE5;
        qWarning() << "Detected a KDE5 session, but we don't have Status Notifier support. Disabling the systray icon";
    }
```

and the `SystrayBackendType::KDE5` value can be removed from the enum too.

agilob commented:
Hah interesting, I'll take a look at it.

Isn't is much better now? ;)

zetok commented:
@agilob btw, merge my PR to this branch? (with `--ff`, so that there'll be no useless merge commit)

agilob commented:
When I fix the issue above I'll make a new branch from this diff to avoid having merge conflicts again...

agilob committed:
remove unused code, better logging for systray, removed KDE5 from enum

agilob committed:
change default quit icon in tray menu

agilob commented:
Could you try this? If this doesn't work I'll remove the icon. I suppose this is because there is style for onHover in .css and your icon is changed from KDE5 to GTK on mouse hover.

zetok commented:
@agilob err, you still got unneeded code there.. (which wouldn't be the case if you did push my commit)

agilob commented:
what code is unused? what commit?

zetok commented:
https://github.com/agilob/qTox/pull/2

agilob commented:
pls ;_;

agilob commented:
Closing because it's merged o/
